stl/inc/limits

```diff
@@ -1009,24 +1011,103 @@ public:
+    return static_cast<int>(_Nn) - static_cast<int>(_Val);
+}
+
+// Implementation of countr_zero without using specialized CPU instructions.
+// Used at compile time and when said instructions are not supported.
+// see "Hacker's Delight" section 5-4
+template <class _Ty>
+_NODISCARD constexpr int _Countr_zero_fallback(const _Ty _Val) noexcept {
+    constexpr int _Digits = std::numeric_limits<_Ty>::digits;
+    return _Digits - _Countl_zero_fallback(static_cast<_Ty>(static_cast<_Ty>(~_Val) & static_cast<_Ty>(_Val - 1)));
+}
+
+#if defined(_M_IX86) || defined(_M_X64)
+
+// TRANSITION, VS 2019 16.8 Preview 1, intrin0.h will declare _tzcnt*
+extern "C" {
+extern int __isa_available;
+#ifdef __clang__
+#define _TZCNT_U32 __builtin_ia32_tzcnt_u32
+#define _TZCNT_U64 __builtin_ia32_tzcnt_u64
+#else // ^^^ __clang__ / !__clang__ vvv
+__MACHINEX86_X64(unsigned int _tzcnt_u32(unsigned int))
+__MACHINEX64(unsigned __int64 _tzcnt_u64(unsigned __int64));
+#define _TZCNT_U32 _tzcnt_u32
+#define _TZCNT_U64 _tzcnt_u64
+#endif // __clang__
+}
+
+template <class _Ty>
+_NODISCARD int _Checked_x86_x64_countr_zero(const _Ty _Val) noexcept {
+
+    constexpr int _Digits = numeric_limits<_Ty>::digits;
+    constexpr _Ty _Max    = (numeric_limits<_Ty>::max)();
+
+#ifndef __AVX2__
+    const bool _Have_tzcnt = __isa_available >= __ISA_AVAILABLE_AVX2;
+    if (!_Have_tzcnt && _Val == 0) {
+        return _Digits;
+    }
+#endif // __AVX2__
+
+    if constexpr (_Digits <= 32) {
+        // Intended widening to int. This operation means that a narrow 0 will widen
+        // to 0xFFFF....FFFF0... instead of 0. We need this to avoid counting all the zeros
+        // of the wider type.
+        return static_cast<int>(_TZCNT_U32(static_cast<unsigned int>(~_Max | _Val)));
+    } else {
+#ifdef _M_IX86
+        const unsigned int _High = _Val >> 32;
+        const unsigned int _Low  = static_cast<unsigned int>(_Val);
+        int _Result              = _Checked_x86_x64_countr_zero(_Low);
+        _Result += _Low == 0 ? _Checked_x86_x64_countr_zero(_High) : 0;
+        return _Result;
+#else // ^^^ _M_IX86 / !_M_IX86 vvv
+        return static_cast<int>(_TZCNT_U64(_Val));
+#endif // _M_IX86
+    }
+}
+#undef _TZCNT_U32
+#undef _TZCNT_U64
+#endif // defined(_M_IX86) || defined(_M_X64)
+
+template <class _Ty>
+constexpr bool _Is_standard_unsigned_integer =
     _Is_any_of_v<remove_cv_t<_Ty>, unsigned char, unsigned short, unsigned int, unsigned long, unsigned long long>;
 
 template <class _Ty, enable_if_t<_Is_standard_unsigned_integer<_Ty>, int> = 0>
 _NODISCARD constexpr int _Countr_zero(const _Ty _Val) noexcept {
-    if (_Val == 0) {
-        return numeric_limits<_Ty>::digits;
-    }
-
-    if constexpr (sizeof(_Ty) <= sizeof(unsigned int)) {
-        return __builtin_ctz(_Val);
+#ifdef __cpp_lib_is_constant_evaluated
+    if (_STD is_constant_evaluated()) {
+#if defined(_M_IX86) || defined(_M_X64)
+        return _Checked_x86_x64_countr_zero(_Val);

```

Comment:
This is causing the tests to fail. It's saying "for constant evaluation, call the x86/x64 intrinsic-optimized version" which isn't `constexpr`. The affected test directories are:

* `std\tests\P0553R4_bit_rotating_and_counting_functions`
* `libcxx\test\std\numerics\bit\bitops.count`
* `libcxx\test\std\numerics\numeric.ops\numeric.ops.gcd`
* `libcxx\test\std\numerics\numeric.ops\numeric.ops.lcm`